docs/proposals/20200602-machine-phase-hooks.md

```diff
@@ -96,6 +103,7 @@ drained and/or associated instance terminated.
 
 ### Goals
 
+- Define an initial set of hook points for the provision phase.
```

Comment:
I think that the mechanism defined in https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220221-runtime-SDK.md provides a better semantic for the integration with external components. Have you considered this idea?

Reply:
Thanks, I will review that proposal to see how it might fit here.

Reply:
Just reviewing the runtime-sdk proposal again and I think in the future it will allow this use case, assuming integration points of `preProvision`, `preDrain`, `preTerminate` for the Machine lifecycle are implemented. Is there any progress towards a list of hooks that will actually be implemented?

One thing that does seem a downgrade from the machine lifecycle hooks is that there is a lower observability with the runtime hooks. With a machine lifecycle hook an admin/controller has made a decision to add the hook so that they can intervene with the lifecycle. If I fetch the machine object I can see the hook is there and so as a naive user, why isn't my Machine draining, oh, there's a hook. I don't see mention of how this will be handled for runtime hooks as there won't be anything stored on the machine to show what the state is.

Another issue, is that machine deletion hooks allow for manual intervention from a user. They can be overriden when there's an issue by and admin removing the hook, they can also be manually added by an admin if they want to for some reason, neither of those cases seem to be covered by runtime hooks which could cause issues down the line, a get out of jail free card seems like a must right?

Reply:
> Is there any progress towards a list of hooks that will actually be implemented?

No, this is up to the community to decide, but I see preProvision, preDrain, preTerminate non controversial

> One thing that does seem a downgrade from the machine lifecycle hooks is that there is a lower observability with the runtime hooks
> Another issue, is that machine deletion hooks allow for manual intervention from a user. 

Nothing prevents doing the same with lifecycle hooks, e.g making the hooks work conditionally given a set of annotations. 

But this isn't always what you want, imaging a step that MUST be performed otherwise the heath of your cluster is compromised, most probably you don't want the user to be able to skip it with just a kubectl edit. 

With runtime hooks you can do both + we usually are surfacing with conditions messages that come back from the hook call, which instead is missing with old styled hooks

Reply:
> But this isn't always what you want, imaging a step that MUST be performed otherwise the health of your cluster is compromised, most probably you don't want the user to be able to skip it with just a kubectl edit.

I'm not sure that that is something that we can and should prevent. In general in Kubernetes the pattern seems to be that when things are getting stuck folks can intervene. Some examples:
* Removing finalizers on disks
* Removing finalizers on Cluster API resources (e.g. Machines)
* Deleting ExtensionConfigs
* ...

In general I like the thought of making the hooks that should be called visible on the corresponding object (Cluster, Machine, ...)